File: hungries_backend/src/auth/usecase/signUpUseCase.ts
import IAuthRepositry from "../domain/IAuthRepository";
import IPasswordService from "../service/IPasswordService";

export default class SignUpUseCase{

    constructor(
        private authRepository:IAuthRepositry,
        private passwordService:IPasswordService)
    {}

    public async execute(
        name:string,
        email:string,
        password:string,
        authType:string
    ):Promise<string>{
        
        const user=await this.authRepository.find(email).then((user)=>{
            return Promise.reject('User already Exist')
        }).catch(async (err)=>{if(err=='User not found'){
            
        
        
        let passwordHash
        if(password)
        passwordHash=await this.passwordService.hash(password)
        else passwordHash=undefined

        const userId=this.authRepository.add(
            email,
            name,
            authType,
            passwordHash,
        )
        
        return userId
        }})

        return ''
        
    
    }
}

File: hungries/auth/lib/src/usecases/signup_usecase.dart
import 'package:auth/src/domain/signUp_service.dart';

import '../domain/token.dart';
import 'package:async/async.dart';


class SignUpUsecase {
  final ISignUpService _iSignUpService;

  SignUpUsecase(this._iSignUpService);

  Future<Result<Token>> execute(
      String name, String email, String password) async {
    return await _iSignUpService.signup(name, email, password);
  }
}
